enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,175 @@
+
+### Implementation Details/Notes/Constraints [optional]
+
+The clusterVersion object has several conditions in its status. Example:
+
+```yaml
+  status:
+    conditions:
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Done applying 4.9.0-0.nightly-2021-07-05-092650
+      status: "True"
+      type: Available
+    - lastTransitionTime: "2021-07-06T12:03:02Z"
+      status: "False"
+      type: Failing
+    - lastTransitionTime: "2021-07-05T20:53:02Z"
+      message: Cluster version is 4.9.0-0.nightly-2021-07-05-092650
+      status: "False"
+      type: Progressing
+    - lastTransitionTime: "2021-07-05T20:05:15Z"
+      message: 'Unable to retrieve available updates: currently reconciling cluster
+        version 4.9.0-0.nightly-2021-07-05-092650 not found in the "stable-4.8" channel'
+      reason: VersionNotFound
+      status: "False"
+      type: RetrievedUpdates
+    ...
+```
+
+The machineHealthCheck supports pausing by setting an annotation on it:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: example
+  namespace: openshift-machine-api
+  annotations:
+    cluster.x-k8s.io/paused: ""
+  ...
+```
+
+The new controller will add the pause annotation on the machineHealthChecks when the clusterVersion's `Progressing`
+condition switches to `"True"`. When that condition switches back to `"False"`, the controller will remove the
+annotation.
+
+The new controller will be implemented in the `openshift/machine-api-operator` repository, and be deployed in the
+`machine-api-controllers` pod, both alongside the machine-healthcheck-controller.
+
+### Risks and Mitigations
+
+Nothing we are aware of today
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Should admins be able to enable / disable this feature? (IMHO no, I don't see a usecase when this has disadvantages)
+
+2. Today there is no way to coordinate multiple users of the pause feature of machineHealthChecks. The value of the
+annotation is supposed to be empty. This means that the annotation might already be set by someone else when the new
+controller want to set it as well. That introduces the question what to do in this case after cluster upgrade: remove
+the annotation, or keep it? In the latter case: how to keep track whether the controller set the annotation or not (e.g.
+in another annotation?)
+
+### Test Plan
+
+TODO (need to check existing e2e tests, and if we extend them with a test case for this, and how much extra load to CI
+it would add)
+
+### Graduation Criteria
+
+TODO I have no good idea yet
+- if this is alpha, beta or stable (do we need this at all, it doesn't introduce a new API)?
+- if this is Dev Preview, Tech Preview or GA?
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+The new controller lives in the machine-api-operator image so the upgrades will be driven by the CVO which will fetch
```

Comment:
is there any special thing needs to be done when the upgraded node is the one that runs this new controller?
e.g. if it's the first node to be upgraded, there's a race between this controller to pause all MHCs, and whatever does the upgrade to reboot the node.
If a reboot takes place before this controller paused all MHCs, and other node was upgraded at the same time (is it possible?), then once MHC comes up, together with this new controller, there's another race, and MHC might remediate the other node before this new controller was able to pause all MHCs.

Maybe I've gone too far :)

Reply:
I can't follow tbh... why does it make a difference if the new controller runs on an updated node or an old one? What kind of race is there? 🤔 Maybe remediation isn't paused in time for all nodes during the upgrade which introduces this new controller, but is that a problem?

Reply:
I'll try to describe the flow (which is probably a super corner case, it's just an example):

1. MHC CR created by an admin
2. upgrade is starting
3. the node that runs the `machine-api-controllers` pod (node A) is the first to be upgraded, and is rebooted before the proposed controller had a chance to pause all MHCs
4. `machine-api-controllers` pod  is scheduled on other node (node B)
5. MHC (the controller) comes up (on node B) before the proposed controller, and tries to remediate node A even though it's down for upgrade

Reply:
I don't think there is anything we can do about it. I also think it isn't an issue, because the same can happen today already? 🤔